File: code/c/15-counting_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

/**
 * Function to perform Counting Sort on an array.
 * Assumes that all elements are non-negative integers.
 *
 * @param arr The array to sort.
 * @param n The size of the array.
 * @param max_value The maximum possible value in the array.
 */
void counting_sort(int arr[], int n, int max_value) {
  int *count = (int *)calloc(max_value + 1, sizeof(int));
  int *output = (int *)malloc(n * sizeof(int));

  // Store the count of each element
  for (int i = 0; i < n; i++) count[arr[i]]++;

  // Change count[i] so that count[i] now contains actual position of this
  // element in output array
  for (int i = 1; i <= max_value; i++) count[i] += count[i - 1];

  // Build the output array
  for (int i = n - 1; i >= 0; i--) {
    output[count[arr[i]] - 1] = arr[i];
    count[arr[i]]--;
  }

  // Copy the output array to arr[]
  memcpy(arr, output, n * sizeof(int));

  // Free allocated memory
  free(count);
  free(output);
}

/**
 * Main function to demonstrate Counting Sort and measure execution time.
 */
int main() {
  int N_values[] = {100000, 200000, 400000, 800000};
  int num_tests = sizeof(N_values) / sizeof(N_values[0]);
  int max_value = 1000000;  // Define a maximum value for elements in the array

  for (int t = 0; t < num_tests; t++) {
    int N = N_values[t];
    int *arr = (int *)malloc(N * sizeof(int));

    // Seed the random number generator
    srand(123);

    // Fill the array with random integers between 0 and max_value
    for (int i = 0; i < N; i++) {
      arr[i] = rand() % (max_value + 1);
    }

    // Measure the start time
    clock_t start_time = clock();

    // Perform Counting Sort
    counting_sort(arr, N, max_value);

    // Measure the end time
    clock_t end_time = clock();

    // Calculate the elapsed time in seconds
    double time_spent = (double)(end_time - start_time) / CLOCKS_PER_SEC;

    // Verify that the array is sorted
    int is_sorted = 1;
    for (int i = 1; i < N; i++) {
      if (arr[i - 1] > arr[i]) {
        is_sorted = 0;
        break;
      }
    }

    // Print the result and time
    printf("Counting Sort with N = %d\n", N);
    if (is_sorted)
      printf("Array is sorted.\n");
    else
      printf("Array is NOT sorted.\n");
    printf("Time taken: %f seconds\n\n", time_spent);

    // Free the allocated memory
    free(arr);
  }

  return 0;
}

File: code/zig/15-counting_sort.zig
const std = @import("std");
const print = std.debug.print;
const time = std.time;
const Instant = time.Instant;
const Allocator = std.mem.Allocator;
const allocator = std.heap.page_allocator;
const rand = std.rand;

/// Function to perform Counting Sort on an array.
///
/// # Parameters
///
/// - `allocator`: The allocator to use for temporary arrays.
/// - `arr`: The array to sort.
/// - `max_value`: The maximum possible value in the array.
fn countingSort(alloc: Allocator, arr: []usize, max_value: usize) !void {
    const n = arr.len;

    // Create count array with size max_value + 1, initialized to zero
    var count = try alloc.alloc(usize, max_value + 1);
    defer alloc.free(count);
    @memset(count, 0);

    // Create output array
    var output = try alloc.alloc(usize, n);
    defer alloc.free(output);

    // Store the count of each element
    for (arr) |value| {
        count[value] += 1;
    }

    // Change count[i] so that count[i] now contains actual position
    // of this element in output array
    var total: usize = 0;
    for (0..max_value + 1) |i| {
        const old_count = count[i];
        count[i] = total;
        total += old_count;
    }

    // Build the output array
    for (arr) |value| {
        const index = count[value];
        output[index] = value;
        count[value] += 1;
    }

    // Copy the output array back to arr[]
    @memcpy(arr, output);
}

/// Function to verify if an array is sorted.
///
/// # Parameters
///
/// - `arr`: The array to check.
fn isSorted(arr: []usize) bool {
    const n = arr.len;
    for (0..n - 1) |i| {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

/// Main function to demonstrate Counting Sort and measure execution time.
pub fn main() !void {
    const N_values = [_]usize{ 10_000, 20_000, 30_000, 40_000 };

    // Seed the random number generator
    var rng = rand.DefaultPrng.init(123); // Xoshiro256 is good enough

    for (N_values) |N| {
        // Allocate memory for the array
        var arr = try allocator.alloc(usize, N);
        defer allocator.free(arr);

        // Fill the array with random integers
        for (0..N) |i| {
            const random_value: usize = rng.next() % N;
            arr[i] = random_value;
        }

        // Measure the start time
        const start = try Instant.now();

        // Perform Counting Sort
        const max_value: usize = 1_000_000;
        try countingSort(allocator, arr, max_value);

        // Measure the end time
        const end = try Instant.now();

        // Calculate the elapsed time in seconds
        const elapsed: f64 = @floatFromInt(end.since(start));
        const time_spent = elapsed / time.ns_per_s;

        // Verify that the array is sorted
        const is_sorted = isSorted(arr);

        // Print the result and time
        print("Insertion Sort with N = {d}\n", .{N});
        if (is_sorted) {
            print("Array is sorted.\n", .{});
        } else {
            print("Array is NOT sorted.\n", .{});
        }
        print("Time taken: {d} seconds\n\n", .{time_spent});
    }
}
